pkg/k8s/provisioner/provisioner_controller.go

```diff
@@ -0,0 +1,128 @@
+package provisioner
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/operator-framework/rukpak/api/v1alpha1"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	apiresource "k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+func (r *Reconciler) ReconcileProvisionerClass(ctx context.Context, req ctrl.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("request", req)
+	log.V(1).Info("reconciling provisionerclass")
+
+	pc := &v1alpha1.ProvisionerClass{}
+	if err := r.Client.Get(ctx, req.NamespacedName, pc); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	bundles := &v1alpha1.BundleList{}
+	if err := r.Client.List(ctx, bundles, client.InNamespace(req.Namespace)); err != nil {
+		log.Error(err, "failed to list all the bundles in the req.Namespace")
+		return ctrl.Result{}, err
+	}
+
+	filtered := []*v1alpha1.Bundle{}
+	for _, b := range bundles.Items {
+		log.Info("provisioner", "processing bundle name", b.Name)
+		if string(b.Spec.Class) != pc.Name {
+			log.Info("provisioner", "found bundle name that does not reference the current provisioner class", b.Name)
+			continue
+		}
+		filtered = append(filtered, &b)
+	}
+	if len(filtered) == 0 {
+		log.Info("no bundles found specifying the current provisionerclass")
+		return ctrl.Result{}, nil
+	}
+
+	var errors []error
+	for _, bundle := range filtered {
+		log.Info("provisioner", "found bundle name that references the current provisionerclass", bundle.Name)
+		if err := r.ensureBundleVolume(ctx, bundle.GetName()); err != nil {
+			errors = append(errors, err)
+		}
+	}
+
+	return ctrl.Result{}, utilerrors.NewAggregate(errors)
+}
+
+// ensureBundleVolume is a helper method responsible for ensuring the bundle's status
+// references an existing volume that will the content stored in the bundle sources'
+// filesystem.
+func (r *Reconciler) ensureBundleVolume(ctx context.Context, bundleName string) error {
+	fresh := &v1alpha1.Bundle{}
+	if err := r.Client.Get(ctx, getNonNamespacedName(bundleName), fresh); err != nil {
```

Comment:
Is the reason we strip the namespace from the bundle here because each bundle name is unique across the cluster?

Reply:
I can see that that is the case based on the cluster-scoped Bundle CRD. I am a little surprised, I know we wanted the Instance CRD to be cluster scoped but maybe namespacing bundles could help with storage and tenancy.

For example, as a user in one namespace, I want to install a Bundle. I won't be able to see Bundles in other namespaces necessarily. When I go to create a Instance from the Bundle, that is cluster-scoped and everyone can see it. Other users installing the same bundle will not be able to create an Instance from the same Bundle since one already exists. But they could install it onto the cluster in a separate mountpath on the PV and use it after the other Instance is removed, for example.

Reply:
I think these are all good point to bring up. In the context of this work, where we haven't built out an Instance API yet, there's no real clear delineation of responsibilities of a Bundle vs. Instance resource as they're muddled together and that logic lives in the Bundle controller currently. As a result of that, the Bundle is cluster-scoped, adn we essentially utilize a global `rukpak` namespace, mainly for ease of local development as we don't have to care about introducing proper RBAC, or worrying about how we'll access resources defined in another namespace. 

This structure is highly likely to change, and outside of the rukpak CRDs, the overall structure is likely contingent on the individual provisioner until we get a better idea of how to consolidate some of these components into a re-usable library to avoid duplicate of effort on a per-provisioner-case.

I still need to continue working out in my head all of the various end-user workflow(s), and haven't gotten a chance to start thinking about how we'd handle authorization/activation of content referenced in individual Bundles (or a series of Bundles that share a label selector), or how a provisioner would create those resources. It's likely that at least one of these APIs will transition to a namespaced-scoped resource.